### PR TITLE
Add Namespaces setting

### PR DESCRIPTION
**Pull request by orivej**
This option holds a comma-separated list of paths bound to Linux network
namespaces to enter with setns() before starting the user session.

The main intention is to let user sessions start in a protected network
namespace, while the display server remains in the initial network namespace to
talk to udev, and udev remains in the initial network namespace to talk to
systemd.

[ChangeLog][General] Add "Namespaces" setting under "General" to enter specified
Linux namespaces before starting the user session

---

In my case, user sessions run in a network namespace with only OpenVPN-created network interfaces for consistent quality of the internet, sidestepping arbitrary deficiencies of various internet provides I happen to use and ensuring that user applications will not connect to the internet directly while OpenVPN connection is down (e.g. it is reconnecting).

To check this functionality, either run `sudo ip netns add mynet` and configure sddm with
```
[General]
Namespaces=/run/netns/mynet
```
or use `Namespaces=/proc/1/ns/net` as a no-op.

### CODE REVIEW

**Comment by orivej** [comment]
I have rebased this on the current `develop`. I have been using sddm with this patch all this time (my setup would not work without it), but it seems there is no demand from other users…

File: src/helper/UserSession.cpp [hunk 1]
```diff
@@ -110,6 +111,22 @@ namespace SDDM {
             }
         }
 
+        // enter Linux namespaces
```
Comment (plfiorini) [inline src/helper/UserSession.cpp]: this block of code is linux specific, should be wrapped in a `#ifdef Q_OS_LINUX`

File: src/helper/UserSession.cpp [hunk 2]
```diff
@@ -110,6 +111,22 @@ namespace SDDM {
...
+        for (const QString &ns: mainConfig.Namespaces.get()) {
+            qInfo() << "Entering namespace" << ns;
+            int fd = ::open(qPrintable(ns), O_RDONLY);
+            if (fd < 0) {
+                qCritical("open(%s) failed: %s", qPrintable(ns), strerror(errno));
+                exit(Auth::HELPER_OTHER_ERROR);
+            }
+            if (setns(fd, 0) != 0) {
```
Comment (plfiorini) [inline src/helper/UserSession.cpp]: not an expert of namespaces but shouldn't the second argument be CLONE_NEWNET?

Comment (orivej) [reply src/helper/UserSession.cpp]: Indeed, I documented that `Namespaces=` joins network namespaces, but wrote the code that allows to join any kind of namespace. I do not see any benefits from this limitation, so I'd rather update the documentation.

File: data/man/sddm.conf.rst.in [hunk 3]
```diff
@@ -58,6 +58,13 @@ OPTIONS
         or "compose" for dead keys support.
         Leave this empty if unsure.
 
+`Namespaces=`
+	Comma-separated list of paths bound to Linux network namespaces to enter
+	with setns() before starting the user session.  For example, to enter
+	network namespace `mynet` created with `ip netns add mynet`, the value
+	might be `/run/netns/mynet`.
+	Default value is empty.
```
Comment (plfiorini) [inline data/man/sddm.conf.rst.in]: Add something along the line of "This setting is relevant only on Linux, it is completely ignored on other operating systems".

I know it is redundant but it's clearer

File: src/helper/UserSession.cpp (hunk 1 above)
Comment (orivej) [reply src/helper/UserSession.cpp]: Does SDDM support or plans to support other OSs?

Comment (plfiorini) [reply src/helper/UserSession.cpp]: FreeBSD users might be interested in using SDDM but personally I'm not working on that goal.

**Comment by orivej** [comment]
Updated.

**Comment by orivej** [comment]
Thank you!

**Pull request merged**